### ISSUE TITLE
Gulp and browser-sync in "gulp watch-bs" task.

### ISSUE THREAD

**Issue by patxipierce** [understrap/understrap#331]
Hello,

I was having trouble with the `gulp watch-bs` command. The browser would sync before the files were compiled resulting in no css updates. In some cases they would fail to update at all. And it turned out the order of the commands matters.

I changed the order and now it works fine:

`gulp.task('watch-bs', ['watch', 'scripts', 'browser-sync'], function () { });`

Instead of:

`gulp.task('watch-bs', ['browser-sync', 'watch', 'scripts'], function () { });`

**Comment by robwakeman** [comment]
I was experiencing something similar too.

However, I think it was only intermittent, and maybe it depended on the order in which I was editing theme files (e.g. PHP, then CSS etc). It was hard to tell due to the intermittent nature.

I’m using the understrap-child theme
https://github.com/holger1411/understrap-child


I had made a couple of small amendments to understrap-child/gulpfile.js as follows:

To watch for php file changes in all sub-directories, to match the same paths as in understrap/gulpfile.js:

```
var browserSyncWatchFiles = [
    './css/*.min.css',
    './js/*.min.js',
    './**/*.php'
];
```

and I had added my custom js (child-theme-custom.js) to the scripts task:

```
gulp.task('scripts', function() {
    var scripts = [
        basePaths.dev + 'js/tether.js', // Must be loaded before BS4

        // Start - All BS4 stuff
        basePaths.dev + 'js/bootstrap4/bootstrap.js',

        // End - All BS4 stuff

        // RW added
        'js/child-theme-custom.js',

        basePaths.dev + 'js/skip-link-focus-fix.js'
    ];
```

But I can’t see how these amendments could’ve caused the issue, though I’m no gulp expert, so not sure.

Anyway, so far, so good. I’ll report back if I see the issue again.

**Comment by robwakeman** [comment]
I'm revisiting this, after downloading the latest understrap-child files from [https://github.com/understrap/understrap-child](https://github.com/understrap/understrap-child).

The issue is still there i.e. 
If I make an edit in:
themes/understrap-child/sass/theme/_child_theme.scss
the changed CSS is correctly injected into the browser with browser-sync.

But, if I then make an edit in say a php template e.g.
themes/understrap-child/loop-templates/content-rw-page.php
that latest CSS change doesn't display in the browser.

A clunky workaround is to then resave the file:
themes/understrap-child/sass/theme/_child_theme.scss
but that's not a good solution obviously.

This issue occurs whether I use:

`gulp.task('watch-bs', ['browser-sync', 'watch', 'scripts'], function () { });`

or

`gulp.task('watch-bs', ['watch', 'scripts', 'browser-sync'], function () { });`

I tried to add `stream: true` in browserSyncOptions, as suggested [here](https://stackoverflow.com/questions/43395463/how-to-deal-with-browser-cache-while-using-browsersync), but this didn't help.

The only way I've managed to resolve it is with a browser hard refresh to clear the cache (works in Chrome anyway). Then, with any further changes in either _child_theme.scss or one of the php template files, the latest styles are correctly injected into the browser.

There may be a better solution, but that's as far as I've got for now.

**Comment by Thomas-A-Reinert** [comment]
Hi Rob,

I´m not the gulpfile-guru, but there shouldn´t be a caching problem according to my experiences, except you may have any caching-plugin installed or set .css files to be browser-cached in a customized .htaccess file.

It may be helpful though to always have the chrome dev-tools open anyway - that should avoid caching contents then.

**Comment by robwakeman** [comment]
Thanks Thomas

I'm not aware of having a caching plugin installed and the .htaccess just has the standard WordPress rules in it, nothing else I'm aware of.

I tried restarting gulp, then opened chrome dev tools. This didn't solve it. Again, for me, the only thing that worked was a hard browser refresh. Hmm...

Thanks for the tips, though.

**Comment by 0dp** [comment]
@robwakeman are you developing on a local system or remote/hosted system?
What OS/platform are you working on?
what is the output of `<?php phpinfo(); ?>`?
is node, npm, gulp up to date?

**Comment by robwakeman** [comment]
@0dp Thanks for the suggested checks. I'm developing on a local system with MAMP.

OS/Platform
OS X El Capitan 10.11.6 (15G19009)

Initial output of `<?php phpinfo(); ?>`  :

PHP Version 5.6.10
System	Darwin Robs-MacBook-Pro.local 15.6.0 Darwin Kernel Version 15.6.0: Tue Jan 9 20:12:05 PST 2018; root:xnu-3248.73.5~1/RELEASE_X86_64 x86_64

**Node** was on
v7.6.0
I updated
v9.8.0

**npm** was on
v4.1.2
I updated
v5.8.0

**gulp** is on
v3.9.1
(latest)

After all those updates, the issue is still exactly the same.

Just to be clear, I need to hard refresh the browser only once. After that, any subsequent changes in either css or php files show correctly.

**Comment by 0dp** [comment]
Actually with the phpinfo I wanted to see the php modules loaded, but you can output that with phpinfo(INFO_MODULES);

**Comment by robwakeman** [comment]
Ok, sorry. Here's the [output](http://www.robwakemandev.com/test/phpinfo/phpinfo.php) of phpinfo(INFO_MODULES);

**Comment by robwakeman** [comment]
Ok, here's the [output](http://www.robwakemandev.com/test/phpinfo/phpinfo.html) of phpinfo(INFO_MODULES) on my local machine.

**Comment by 0dp** [comment]
It didn't give me any clues but I will keep investigating. (feel free to remove the files. You don't want that one laying around).

Have you tried to make a clean install of WordPress, no plugins, fresh version of UnderStrap, `npm install`, `gulp watch-bs` to verify it's not an isolated incident?

On a different note, you should probably look into updating your php version :)

**Comment by robwakeman** [comment]
Ok thanks anyway. Output file removed.

I've just tried it with PHP v7.0.0 (I did have that installed, but haven't tended to use it, because it hasn't played well with some older plugins on other sites). The issue remains the same.

Did you see my July 18 2017 comment above about adding the extra array element to `browserSyncWatchFiles`? Could that be causing the problem? Have I written that correctly?

My install of WordPress was clean, without plugins. But, yes I will try another one and see what happens - I'll do that when I get a moment.

**Comment by 0dp** [comment]
I was going to ask you if you've tried without the modifications?

We've added a custom-javascript file in src that will get minified and included so you don't need that line anymore. Secondly, it's better to follow the intended declaration so that  ` 'js/child-theme-custom.js',` becomes `paths.dev + '/js/child-theme-custom.js'`

**Comment by robwakeman** [comment]
I've just tried a separate clean install of WordPress and the latest versions of themes understrap and understrap-child

understrap
v0.8.1

understrap-child
v0.5.1

This included the latest commits on 26/03/18

The only things I did were:

- edit  the proxy in to reflect my localhost URL
themes/understrap-child/gulpconfig.json

- copy in a few PHP template files e.g:
themes/understrap-child/page-templates/rw-fullwidthpage.php
themes/understrap-child/loop-templates/content-rw-page.php

- create a page in WP to use this template

This was so that I could test the issue, i.e. edit a PHP template as well as edit the CSS e.g:
themes/understrap-child/sass/theme/_child_theme_variables.scss

I didn't touch gulpfile.js or any other files.

The issue is still there. And, as before, a single hard refresh sorts it out.
So, do you think this could be just some issue connected with browser-sync and my machine? i.e. not specific to understrap or understrap-child?

The addition of custom-javascript.js is great.
Thanks for the advice re `paths.dev`

**Comment by MrkKr** [comment]
I have same problem as @robwakeman

**Comment by 0dp** [comment]
@MrkKr please provide more info by filling out the form below:
```
Please delete anything that does not apply for you!

## Issue is about..
  - [ ] a Bug Report
  - [ ] a Feature Request
  - [ ] a Suggestion
  - [ ] a Question
  - [ ] I need help

## Please tell us about your environment:

  - **WordPress Version:** 4.9.x
  - **OS Version:** [Windows 7 | Windows 8 | Windows 10] | Linux [Distribution / Version] | MacOS [Version] | Android [Version] | iOS [Version]
  - **Browser:** [ all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari ]
  - **Language:** [ all | PHP | SQL | HTML | S/CSS | Javascript | jQuery | ES6 | JSON | XML]
  - **Node.js & NPM Versions:** Node.js [V X.XX] | NPM [V X.XX]


---

# Issue Overview for Bugreports

## Steps to reproduce?

## *Who* is the bug affecting?

## *What* is affected by this bug?

## *When* does this occur?

## *Where* on the platform does it happen?

## *How* do we replicate the issue?

## Are there any related Issues and/or PRs?

## Screenshots / Video

## Expected behavior

# Todos

 * Test
 * Documentation



### Other Comments
```

**Comment by MrkKr** [comment]
Issue: After make changes in scss files browser is not reload
WP: 4.9.5
OS: Win10
Browser: Chrome 67
PHP 7.2.4 (in local)
Node - 8.11.1
Gulp CLI - 2.0.1
NPM  - 5.6.0

I install understrap and all dependencies by 'npm install'.
All works fine apart css (.scss) sync files. If I make changes in .scss file, there are minify properly, but browser is not reloading.
Earlier, I worked on this same project on Ubuntu 17.04 and it was ok.

**Comment by 0dp** [comment]
and you are executing `gulp watch-bs`
Have you also changed `proxy` in gulpconfig.json:3?

**Comment by MrkKr** [comment]
@0dp Yes, I run gulp by watch-bs. I changes proxy in JSON to "proxy": "localhost/NAME/". Everything sync ok (include JS), apart from css files. I have to refresh manually.

**Comment by 0dp** [comment]
gotcha!

It's a really hard problem to wrap my head around; hope you understand.
There is no real consistency in how this issue manifest it self. You and @robwakeman have this issue across different OS and different versions. I myself can't reproduce the issue at all.

Could you try and copy this `gulpfile.js` into your project and see if anything changes?

**Comment by MrkKr** [comment]
@0dp no problem, I understand, that its strange.
gulpfile.js code:
`// Defining requirements
var gulp = require( 'gulp' );
var plumber = require( 'gulp-plumber' );
var sass = require( 'gulp-sass' );
var watch = require( 'gulp-watch' );
var cssnano = require( 'gulp-cssnano' );
var rename = require( 'gulp-rename' );
var concat = require( 'gulp-concat' );
var uglify = require( 'gulp-uglify' );
var merge2 = require( 'merge2' );
var imagemin = require( 'gulp-imagemin' );
var ignore = require( 'gulp-ignore' );
var rimraf = require( 'gulp-rimraf' );
var clone = require( 'gulp-clone' );
var merge = require( 'gulp-merge' );
var sourcemaps = require( 'gulp-sourcemaps' );
var browserSync = require( 'browser-sync' ).create();
var del = require( 'del' );
var cleanCSS = require( 'gulp-clean-css' );
var gulpSequence = require( 'gulp-sequence' );
var replace = require( 'gulp-replace' );
var autoprefixer = require( 'gulp-autoprefixer' );

// Configuration file to keep your code DRY
var cfg = require( './gulpconfig.json' );
var paths = cfg.paths;

gulp.task( 'watch-scss', ['browser-sync'], function() {
    gulp.watch( paths.sass + '/**/*.scss', ['scss-for-dev'] );
});

// Run:
// gulp sass
// Compiles SCSS files in CSS
gulp.task( 'sass', function() {
    var stream = gulp.src( paths.sass + '/*.scss' )
        .pipe( plumber( {
            errorHandler: function( err ) {
                console.log( err );
                this.emit( 'end' );
            }
        } ) )
        .pipe(sourcemaps.init({loadMaps: true}))
        .pipe( sass( { errLogToConsole: true } ) )
        .pipe( autoprefixer( 'last 2 versions' ) )
        .pipe(sourcemaps.write(undefined, { sourceRoot: null }))
        .pipe( gulp.dest( paths.css ) )
    return stream;
});

// Run:
// gulp watch
// Starts watcher. Watcher runs gulp sass task on changes
gulp.task( 'watch', function() {
    gulp.watch( paths.sass + '/**/*.scss', ['styles'] );
    gulp.watch( [paths.dev + '/js/**/*.js', 'js/**/*.js', '!js/theme.js', '!js/theme.min.js'], ['scripts'] );

    //Inside the watch task.
    gulp.watch( paths.imgsrc + '/**', ['imagemin-watch'] );
});

/**
 * Ensures the 'imagemin' task is complete before reloading browsers
 * @verbose
 */
gulp.task( 'imagemin-watch', ['imagemin'], function( ) {
  browserSync.reload();
});

// Run:
// gulp imagemin
// Running image optimizing task
gulp.task( 'imagemin', function() {
    gulp.src( paths.imgsrc + '/**' )
    .pipe( imagemin() )
    .pipe( gulp.dest( paths.img ) );
});

// Run:
// gulp cssnano
// Minifies CSS files
gulp.task( 'cssnano', function() {
  return gulp.src( paths.css + '/theme.css' )
    .pipe( sourcemaps.init( { loadMaps: true } ) )
    .pipe( plumber( {
            errorHandler: function( err ) {
                console.log( err );
                this.emit( 'end' );
            }
        } ) )
    .pipe( rename( { suffix: '.min' } ) )
    .pipe( cssnano( { discardComments: { removeAll: true } } ) )
    .pipe( sourcemaps.write( './' ) )
    .pipe( gulp.dest( paths.css ) );
});

gulp.task( 'minifycss', function() {
  return gulp.src( paths.css + '/theme.css' )
  .pipe( sourcemaps.init( { loadMaps: true } ) )
    .pipe( cleanCSS( { compatibility: '*' } ) )
    .pipe( plumber( {
            errorHandler: function( err ) {
                console.log( err ) ;
                this.emit( 'end' );
            }
        } ) )
    .pipe( rename( { suffix: '.min' } ) )
     .pipe( sourcemaps.write( './' ) )
    .pipe( gulp.dest( paths.css ) );
});

gulp.task( 'cleancss', function() {
  return gulp.src( paths.css + '/*.min.css', { read: false } ) // Much faster
    .pipe( ignore( 'theme.css' ) )
    .pipe( rimraf() );
});

gulp.task( 'styles', function( callback ) {
    gulpSequence( 'sass', 'minifycss' )( callback );
} );

// Run:
// gulp browser-sync
// Starts browser-sync task for starting the server.
gulp.task( 'browser-sync', function() {
    browserSync.init( cfg.browserSyncWatchFiles, cfg.browserSyncOptions );
} );

// Run:
// gulp watch-bs
// Starts watcher with browser-sync. Browser-sync reloads page automatically on your browser
gulp.task( 'watch-bs', ['browser-sync', 'watch', 'scripts'], function() { 
} );

// Run: 
// gulp scripts. 
// Uglifies and concat all JS files into one
gulp.task( 'scripts', function() {
    var scripts = [

        // Start - All BS4 stuff
        paths.dev + '/js/bootstrap4/bootstrap.js',

        // End - All BS4 stuff

        paths.dev + '/js/skip-link-focus-fix.js',

        // Adding currently empty javascript file to add on for your own themes´ customizations
        // Please add any customizations to this .js file only!
        paths.dev + '/js/custom-javascript.js'
    ];
  gulp.src( scripts )
    .pipe( concat( 'theme.min.js' ) )
    .pipe( uglify() )
    .pipe( gulp.dest( paths.js ) );

  gulp.src( scripts )
    .pipe( concat( 'theme.js' ) )
    .pipe( gulp.dest( paths.js ) );
});

// Deleting any file inside the /src folder
gulp.task( 'clean-source', function() {
  return del( ['src/**/*'] );
});

// Run:
// gulp copy-assets.
// Copy all needed dependency assets files from bower_component assets to themes /js, /scss and /fonts folder. Run this task after bower install or bower update

////////////////// All Bootstrap SASS  Assets /////////////////////////
gulp.task( 'copy-assets', function() {

////////////////// All Bootstrap 4 Assets /////////////////////////
// Copy all JS files
    var stream = gulp.src( paths.node + 'bootstrap/dist/js/**/*.js' )
        .pipe( gulp.dest( paths.dev + '/js/bootstrap4' ) );

// Copy all Bootstrap SCSS files
    gulp.src( paths.node + 'bootstrap/scss/**/*.scss' )
        .pipe( gulp.dest( paths.dev + '/sass/bootstrap4' ) );

////////////////// End Bootstrap 4 Assets /////////////////////////

// Copy all Font Awesome Fonts
    gulp.src( paths.node + 'font-awesome/fonts/**/*.{ttf,woff,woff2,eot,svg}' )
        .pipe( gulp.dest( './fonts' ) );

// Copy all Font Awesome SCSS files
    gulp.src( paths.node + 'font-awesome/scss/*.scss' )
        .pipe( gulp.dest( paths.dev + '/sass/fontawesome' ) );

// _s SCSS files
    gulp.src( paths.node + 'undescores-for-npm/sass/media/*.scss' )
        .pipe( gulp.dest( paths.dev + '/sass/underscores' ) );

// _s JS files into /src/js
    gulp.src( paths.node + 'undescores-for-npm/js/skip-link-focus-fix.js' )
        .pipe( gulp.dest( paths.dev + '/js' ) );

// Copy Popper JS files
    gulp.src( paths.node + 'popper.js/dist/umd/popper.min.js' )
        .pipe( gulp.dest( paths.js + paths.vendor ) );
    gulp.src( paths.node + 'popper.js/dist/umd/popper.js' )
        .pipe( gulp.dest( paths.js + paths.vendor ) );
    return stream;
});

// Deleting the files distributed by the copy-assets task
gulp.task( 'clean-vendor-assets', function() {
  return del( [paths.dev + '/js/bootstrap4/**', paths.dev + '/sass/bootstrap4/**', './fonts/*wesome*.{ttf,woff,woff2,eot,svg}', paths.dev + '/sass/fontawesome/**', paths.dev + '/sass/underscores/**', paths.dev + '/js/skip-link-focus-fix.js', paths.js + '/**/skip-link-focus-fix.js', paths.js + '/**/popper.min.js', paths.js + '/**/popper.js', ( paths.vendor !== ''?( paths.js + paths.vendor + '/**' ):'' )] );
});

// Run
// gulp dist
// Copies the files to the /dist folder for distribution as simple theme
gulp.task( 'dist', ['clean-dist'], function() {
  return gulp.src( ['**/*', '!' + paths.bower, '!' + paths.bower + '/**', '!' + paths.node, '!' + paths.node + '/**', '!' + paths.dev, '!' + paths.dev + '/**', '!' + paths.dist, '!' + paths.dist + '/**', '!' + paths.distprod, '!' + paths.distprod + '/**', '!' + paths.sass, '!' + paths.sass + '/**', '!readme.txt', '!readme.md', '!package.json', '!package-lock.json', '!gulpfile.js', '!gulpconfig.json', '!CHANGELOG.md', '!.travis.yml', '!jshintignore',  '!codesniffer.ruleset.xml',  '*'], { 'buffer': false } )
  .pipe( replace( '/js/jquery.slim.min.js', '/js' + paths.vendor + '/jquery.slim.min.js', { 'skipBinary': true } ) )
  .pipe( replace( '/js/popper.min.js', '/js' + paths.vendor + '/popper.min.js', { 'skipBinary': true } ) )
  .pipe( replace( '/js/skip-link-focus-fix.js', '/js' + paths.vendor + '/skip-link-focus-fix.js', { 'skipBinary': true } ) )
    .pipe( gulp.dest( paths.dist ) );
});

// Deleting any file inside the /dist folder
gulp.task( 'clean-dist', function() {
  return del( [paths.dist + '/**'] );
});

// Run
// gulp dist-product
// Copies the files to the /dist-prod folder for distribution as theme with all assets
gulp.task( 'dist-product', ['clean-dist-product'], function() {
  return gulp.src( ['**/*', '!' + paths.bower, '!' + paths.bower + '/**', '!' + paths.node, '!' + paths.node + '/**', '!' + paths.dist, '!' + paths.dist +'/**', '!' + paths.distprod, '!' + paths.distprod + '/**', '*'] )
    .pipe( gulp.dest( paths.distprod ) );
} );

// Deleting any file inside the /dist-product folder
gulp.task( 'clean-dist-product', function() {
  return del( [paths.distprod + '/**'] );
} );
`

**Comment by 0dp** [comment]
@MrkKr So it worked, by changing out the gulp file?

**Comment by MrkKr** [comment]
@0dp no, it still doesn't work.

**Comment by 0dp** [comment]
Can I get you to test on a fresh wordpress with a fresh clone of understrap?
We need to eliminate as many variables as possible

**Comment by MrkKr** [comment]
I am working now on the latest Understrap version. I haven't changed anything in gulpfile.js

**Comment by MrkKr** [comment]
@robwakeman Have you managed with this problem?

**Comment by robwakeman** [comment]
@MrkKr My issue is still as described as in my comment on [Mar 29](https://github.com/understrap/understrap/issues/331#issuecomment-377201288).

I'm not sure if your issue is slightly different from mine. If I make an edit in the SCSS
the changed CSS is correctly injected into the browser. But, if I then make an edit in a PHP template, the latest CSS change is lost. I then have to hard refresh once and it's all fine from then on, whether editing SCSS or PHP files.

Whereas, you have stated the following:
_After make changes in scss files browser is not reload_

**Comment by 0dp** [comment]
Ghost in the fucking machine right about now 😬

**Comment by MrkKr** [comment]
@robwakeman yes, you are right, my issue is not same as you, but similar.

**Comment by 0dp** [comment]
There was nothing that cought my attention sorry

**Comment by Thomas-A-Reinert** [comment]
This issue seems to get a bit diffuse with several issues getting mingled up. I´ll close it for housekeeping therefore.

If the issue persists. please open up another issue using our issue-template.
